README.md
# AngularForms

[![Build Status](https://travis-ci.org/robust-team/angular-forms.svg?branch=master)](https://travis-ci.org/robust-team/angular-forms)
[![codecov](https://codecov.io/gh/robust-team/angular-forms/branch/master/graph/badge.svg)](https://codecov.io/gh/robust-team/angular-forms)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/253b251df47e47998d27404a6746978b)](https://www.codacy.com/app/dudemelo/angular-forms?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=robust-team/angular-forms&amp;utm_campaign=Badge_Grade)
[![NPM Version](https://img.shields.io/npm/v/@robust-team/angular-forms.svg)](http://npmjs.org/package/@robust-team/angular-forms)
[![NPM Download](https://img.shields.io/npm/dm/@robust-team/angular-forms.svg)](http://npmjs.org/package/@robust-team/angular-forms)
[![Issue Stats](http://issuestats.com/github/robust-team/angular-forms/badge/pr?style=flat)](http://issuestats.com/github/robust-team/angular-forms)
[![Issue Stats](http://issuestats.com/github/robust-team/angular-forms/badge/issue?style=flat)](http://issuestats.com/github/robust-team/angular-forms)

A Robust Forms Generator for Angular.

## Table of Contents

- [Requirements](#requirements)
- [Usage](#usage)
- [Groups](#groups)
  - [Fieldset](#fieldset)
  - [DataTable](#datatable)
- [Questions](#questions)
  - [Checkbox](#checkbox)
  - [Radio](#radio)
  - [Select](#select)
  - [Text](#text)
    - [Masks](#masks)
  - [TextArea](#textarea)
- [Validations](#validations)
  - [Email](#email)
  - [Max](#max)
  - [Min](#min)
  - [MaxLength](#maxlength)
  - [MinLength](#minlength)
  - [Pattern](#pattern)
  - [Required](#required)
- [Dependencies](#dependencies)

## Requirements

- **@angular/common**: 4.0.0 or higher,
- **@angular/core**: 4.0.0 or higher,
- **@angular/forms**: 4.0.0 or higher,
- **@fagnerlima/ng-mask**: 2.2.1 or higher,
- **@ngx-translate/core**: 7.0.0 or higher,
- **rxjs**: 5.4.2 or higher.

## Usage

1. Install AngularForms using npm:

```
npm i @robust-team/angular-forms
```

2. Import the styles.css:

```
node_modules/@robust-team/angular-forms/assets/css/styles.css
```

3. Import the AngularFormsModule into Module class.

```typescript
import { AngularFormsModule } from '@robust-team/angular-forms';

@NgModule({
  imports: [
    //...
    AngularFormsModule
  ],
  // ...
})
export class MyModule { }
```

4. Insert the AngularFormsComponent into template.

```html
<rb-angular-forms [groups]="customForm"></rb-angular-forms>
```

The **customForm** attribute represents the JSON coming from API. For example:

```json
[
  {
    "code": "G-01",
    "description": "Fieldset Group",
    "type": "group",
    "questions": [
      {
        "name": "Q-0101",
        "description": "Checkbox Question 0101",
        "type": "checkbox",
        "defaultOption": "false",
        "validations": [
          { "type": "pattern", "message": "Required field.", "value": "true" }
        ]
      },
      {
        "name": "Q-0102",
        "description": "Radio Question 0102",
        "type": "radio",
        "options": ["Option 1", "Option 2"],
        "defaultOption": "Option 1",
        "validations": [
          { "type": "required", "message": "Required field." }
        ]
      },
      {
        "name": "Q-0103",
        "description": "Select Question 0103",
        "type": "select",
        "options": ["Option 1", "Option 2"],
        "editableOption": "Option 2",
        "placeholder": "Select",
        "validations": [
          { "type": "required", "message": "Required field." }
        ]
      },
      {
        "name": "Q-0104",
        "description": "Text Question 0104",
        "type": "text",
        "placeholder": "Text Question 0104",
        "validations": [
          { "type": "required", "message": "Required field." },
          { "type": "minlength", "message": "Min Length: 3", "value": 3 },
          { "type": "maxlength", "message": "Max Length: 10", "value": 10 }
        ]
      },
      {
        "name": "Q-0105",
        "description": "Text Question 0105",
        "type": "text",
        "placeholder": "Text Question 0105",
        "validations": [
          { "type": "required", "message": "Required field." },
          { "type": "pattern", "message": "Format: 00000-000", "value": "^\\d{5}-\\d{3}$" }
        ],
        "mask": "00000-000"
      },
      {
        "name": "Q-0106",
        "description": "TextArea Question 0106",
        "type": "textarea",
        "placeholder": "TextArea Question 0106",
        "validations": [
          { "type": "required", "message": "Required field." }
        ],
        "dependencies": [
          { "code": "Q-0103", "criteria": "equals", "expectedAnswer": "Option 1" }
        ]
      }
    ]
  },
  {
    "code": "G-02",
    "description": "DataTable Group",
    "type": "datatable",
    "validations": [
      { "type": "required", "message": "Min number of registers: 1." },
      { "type": "minlength", "message": "Min number of registers: 1.", "value": 1 },
      { "type": "maxlength", "message": "Max number of registers: 3.", "value": 3 }
    ],
    "questions": [
      [
        {
          "name": "Q-0201",
          "description": "Select Question 0201",
          "type": "select",
          "validations": [
            { "type": "required", "message": "Required field." }
          ],
          "options": ["Option 1", "Option 2", "Option 3"],
          "placeholder": "Select"
        },
        {
          "name": "Q-0202",
          "description": "Text Question 0202",
          "type": "text",
          "validations": [
            { "type": "required", "message": "Required field." }
          ],
          "placeholder": "Text Question 0202"
        }
      ],
      [
        {
          "name": "Q-0201",
          "description": "Select Question 0201",
          "type": "select",
          "validations": [
            { "type": "required", "message": "Required field." }
          ],
          "options": ["Option 1", "Option 2", "Option 3"],
          "placeholder": "Select",
          "answer": "Option 2"
        },
        {
          "name": "Q-0202",
          "description": "Text Question 0202",
          "type": "text",
          "validations": [
            { "type": "required", "message": "Required field." }
          ],
          "placeholder": "Text Question 0202",
          "answer": "Answer 0202"
        }
      ]
    ]
  }
]
```

![screen](docs/screens/angular-forms-01.png)

5. Get form values by adding a local variable to the component and calling the **getForm()** method. For example:

```html
<rb-angular-forms #angularForms [groups]="customForm"></rb-angular-forms>
<button (click)="print(angularForms.getForm())">Print Form</button>
```

```typescript
export class MyComponent {

  print(value: any) {
    console.log(value);
  }
}
```

The output of getForm() is an object that contains:

- **valid**: Informs if the forms is valid, that is, if it agrees with all the validations.
- **value**: An object that contains the groups of questions with their respective answers. Groups and questions are represented by their respective code; and each question has its answer.

Example output:

```typescript
{
  valid: false,
  value: {
    "G-01": [
      {
        "Q-101": { "value": "Option 1", "id": null },
        "Q-102": { "value": "123", "id": null }
      }
    ],
    "G-02": {
      "Q-201": { "value": "Option 1", "id": null },
      "Q-202": { "value": "Option 2", "id": null },
      "Q-203": { "value": null, "id": null },
      "Q-204": { "value": null, "id": null }
    }
  }
}
```

**AngularFormsComponent**'s public methods and properties:

| Method | Function |
|-|-|
| **isValid()** | Returns the status of form. |
| **isPristine()** | Returns if the user has not yet changed some value in the form. |
| **isDirty()** | Returns if the user has changed some value in the form. |
| **getAnswersGroups()** | Returns only the answers grouped in yours respective Groups.  |
| **getAnswers()** | Returns only the answers of simple Questions and of DataTables (that is considered a Question).  |
| **getForm()** | Returns an object with two attributes: **valid** (return of **isValid** method) and **value** (return of **getAnswersGroups** method. |
| **submit()** | Send the form, enabling the validation of all questions. |
| **formGroup** | **FormGroup**'s instance of the **ReactiveForms**. |
| **status** | Returns the form's status: **LOADING*, **READY**, **ERROR** or **null**. |
| **getGroupByCode(code)** | Returns a **Group** according with **code**. |

6. The **AngularFormsComponent** contains two output events:

| Output | Function |
|-|-|
| **ready** | When the form's build is finished. |
| **error** | When occurs any error in form's build (throws the *error object*). |

```html
<rb-angular-forms #angularForms [groups]="customForm" (ready)="onReady()" (error)="onError($event)"></rb-angular-forms>
```

7. If you want only to show the answers from JSON, use the **readOnly** input with **true** value (your default value is **false**).

```html
<rb-angular-forms #angularForms [groups]="customForm" [readOnly]="true"></rb-angular-forms>
```

![screen](docs/screens/angular-forms-02.png)

8. For you config the language to be used by AngularForms, use the **lang** input. There are two langs available: **en-US** (default) and **pt-BR**.

```html
<rb-angular-forms #angularForms [groups]="customForm" lang="pt-BR"></rb-angular-forms>
```

## Groups

The **Group** represents a grouping of questions, which can be of two types: **Fieldset** and **DataTable**.

> The **Group** contains the public method **getQuestionByName**, that returns a question according with **name**.

### Fieldset

**Fieldset** is a grouping of questions that represents a simple fieldset of HTML. The value of your type is "**group**".

```json
{
  "code": "G-01",
  "description": "Fieldset Group",
  "type": "group",
  "questions": [
    {
      "name": "Q-0101",
      "description": "Checkbox Question 0101",
      "type": "checkbox",
      "validations": []
    },
    {
      "name": "Q-0102",
      "description": "Select Question 0102",
      "type": "select",
      "options": ["Option 1", "Option 2"],
      "validations": []
    }
  ]
}
```

### DataTable

**DataTable** is a grouping of questions that represents a data table, where you add many answers.

```json
{
  "code": "G-01",
  "description": "DataTable Group",
  "type": "datatable",
  "questions": [
    [
      {
        "name": "Q-0101",
        "description": "Checkbox Question 0101",
        "type": "checkbox",
        "validations": []
      },
      {
        "name": "Q-0102",
        "description": "Select Question 0102",
        "type": "select",
        "options": ["Option 1", "Option 2"],
        "validations": []
      }
    ]
  ],
  "validations": []
}
```

## Questions

**Question** represents a simple question, which can be of five types:

- **checkbox**;
- **radio**;
- **select**;
- **text**;
- **textarea**.

> The **Question** contains the public setter method **description**, that permits alter label of questions.

### Checkbox

**Checkbox** represents a *checkbox* input from HTML.

```json
{
  "name": "Q-01",
  "description": "Checkbox Question",
  "dependencies": [],
  "type": "checkbox",
  "answer": null,
  "validations": [],
  "disabled": false,
  "defaultOption": "false"
}
```

### Radio

**Radio** represents a *radio* input from HTML.

```json
{
  "name": "Q-01",
  "description": "Radio Question",
  "dependencies": [],
  "type": "radio",
  "answer": null,
  "validations": [],
  "disabled": false,
  "options": ["Option 1", "Option 2"],
  "defaultOption": null
}
```

### Select

**Select** represents a *select* input from HTML.

```json
{
  "name": "Q-01",
  "description": "Select Question",
  "dependencies": [],
  "type": "select",
  "answer": null,
  "validations": [],
  "disabled": false,
  "options": ["Option 1", "Option 2"],
  "defaultOption": null,
  "editableOption": "Option 2",
  "placeholder": "Select an option"
}
```

For the **Radio** and **Select** Questions, the *option* attribute can be a *string array* or an *object array* (**Option** Interface).

```json
{
  // ...
  "options": [
    { "value": "op1", "description": "Option 1" },
    { "value": "op2", "description": "Option 2" }
  ]
  // ...
}
```

### Text

**Text** represents a *text* input from HTML.

```json
{
  "name": "Q-01",
  "description": "Text Question",
  "dependencies": [],
  "type": "text",
  "answer": null,
  "validations": [],
  "disabled": false,
  "mask": null,
  "placeholder": null
}
```

#### Masks

**Mask** is an attribute of **Text Question** that uses the patterns of the [NgMask](https://fagnerlima.github.io/ng-mask/) library.

```json
{
  "name": "Q-01",
  "description": "Text Question",
  "dependencies": [],
  "type": "text",
  "answer": null,
  "validations": [],
  "mask": "integer(6)",
  "placeholder": null
}
```

Read more in [https://fagnerlima.github.io/ng-mask/](https://fagnerlima.github.io/ng-mask/).

### TextArea

**TextArea** represents a *text* input from HTML.

```json
{
  "name": "Q-01",
  "description": "Text Question",
  "dependencies": [],
  "type": "textarea",
  "answer": null,
  "validations": [],
  "disabled": false,
  "placeholder": null
}
```

### Answer

**Answer** represents the question's answer.

```json
{
  "value": "My answer",
  "id": 1
}
```

The **value** property represents the user's answer itself, while the **id** property is the answer's *identifier* into *database*, for example.

## Validations

**Validations** can be used in **Question** and **DataTable**, which can be of seven types:

- **email**;
- **max**;
- **min**;
- **maxlength**;
- **minlength**;
- **pattern**;
- **required**.

### Email

E-mail validation.

```json
{
  "type": "email",
  "message": "Email field."
}
```

### Max

Max value validation for numbers.

```json
{
  "type": "max",
  "message": "Max value: 100.",
  "value": 100
}
```

### Min

Min value validation for numbers.

```json
{
  "type": "min",
  "message": "Min value: 10.",
  "value": 10
}
```

### MaxLength

Max length validation.

```json
{
  "type": "maxlength",
  "message": "Max length: 12.",
  "value": 12
}
```

### MinLength

Min length validation.

```json
{
  "type": "minlength",
  "message": "Min length: 6.",
  "value": 6
}
```

### Pattern

Regular expression validation.

```json
{
  "type": "pattern",
  "message": "Pattern: 00000-000.",
  "value": "^\\d{5}-\\d{3}$"
}
```

### Required

Obligatory field validation. The **requiredTrue** field is used for checkbox validation.

```json
{
  "type": "required",
  "message": "Required field.",
  "requiredTrue": false
}
```

## Dependencies

**Dependencies** are Questions that determine if a Question should be shown or no. The criterias can be of four types:

- **equals**;
- **lessthan**;
- **greaterthan**;
- **notequals**.

```json
{
  "name": "Q-0106",
  "description": "Question TextArea 0106",
  "type": "textarea",
  "placeholder": "Question TextArea 0106",
  "validations": [
    { "type": "required", "message": "Required field." }
  ],
  "dependencies": [
    { "code": "Q-0103", "criteria": "equals", "expectedAnswer": "Option 1" },
    { "code": "Q-0104", "criteria": "greaterthan", "expectedAnswer": "10" }
  ]
}
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="3.0.0-beta1"></a>
## [3.0.0-beta1](https://github.com/robust-team/angular-forms/compare/v2.1.0...v3.0.0-beta1) (2017-10-10)

### Changes

* **Answers**' type into **Questions** changed of *string* to **Answer** object.

### Bug Fixes

* Fixed **isPristine** and **isDirty** methods into **AngularFormsComponent**.

<a name="2.1.0"></a>
## [2.1.0](https://github.com/robust-team/angular-forms/compare/v1.2.0...v2.1.0) (2017-09-28)

### Features

* Added in the **Group** model:
  * **setDescription** method;
  * **getQuestionByName** abstract method;
* Added in the **Question** model:
  * **setDescription** method;
* **Status** enum.
* Added in **AngularFormsComponent**:
  * **CSS class** with question name into template's fieldset;
  * **Output** events error and ready;
  * **getGroupByCode** public method;
  * **getStatus** public method;
  * **Errors' catch** for Promises.
* Changed in **DataTableComponent**:
  * **CSS class** with **question name** into template's fieldset;
  * **Output** events error and ready;
  * **getStatus** public method
  * **Errors' catch** for Promises.

### Changes

* Use of **Promises** into AngularForms and ReactiveFormsFactory.
* Changed in the **Group** model:
  * **Protected visibility** in properties;
  * Generic Type **<QuestionListType>**;
  * **question** property;
* Changed in the **Question** model:
  * **Protected visibility** in properties;
* Changed in **AngularFormsComponent**:
  * **Async** load;
  * **ngOnChanges** event for build form;
* Changed in **DataTableComponent**:
  * **Async** load.

### Bug Fixes

* Fixed initialization of answer property in Choice model.

<a name="1.2.0"></a>
## [1.2.0](https://github.com/robust-team/angular-forms/compare/v1.1.0...v1.2.0) (2017-09-22)

### Features

* Addition of **isPristine** and **isDirty** methods.

<a name="1.1.0"></a>
## [1.1.0](https://github.com/robust-team/angular-forms/compare/v1.0.1...v1.1.0) (2017-09-21)

### Features

* Creation of **Option** Interface for the **Choice Questions**.

### Bug Fixes

* Addition of "no registers" message into **Datatable**.

<a name="1.0.1"></a>
## [1.0.1](https://github.com/robust-team/angular-forms/compare/v1.0.0...v1.0.1) (2017-09-01)

### Bug Fixes

* **Editable option** fixed.

<a name="1.0.0"></a>
## [1.0.0](https://github.com/robust-team/angular-forms/compare/v0.5.2...v1.0.0) (2017-09-01)

### Features

* Addition of *asterisk* to required form controls.
* Addition of the **editableOption** attribute in **Select**.
* Addition of the **disabled** attribute in **Questions** into **Fieldset**.

### Changes

* Creation of the **styles.css** file and deletion of the inline styles into the components.
* **String** updated to **StringUtils**.
* Constructors of the **Groups**, **Questions** and **Validations** updated.
* Methods into **DependencyService** updated to static methods.
* **GroupType**, **QuestionType**, **ValidationType** and **DependencyCriteria** Enum created.
* **ValidationFactory** created.

<a name="0.5.2"></a>
## [0.5.2](https://github.com/robust-team/angular-forms/compare/v0.5.1...v0.5.2) (2017-07-19)

### Changes

* Addition of the submit method in the AngularFormsComponent.

<a name="0.5.1"></a>
## [0.5.1](https://github.com/robust-team/angular-forms/compare/v0.5.0...v0.5.1) (2017-07-19)

### Changes

* Refactoring of the methods of get answers for convert boolean answers to string type.

### Bug Fixes

* **Dependencies** fixed when the readOnly option is enabled.
* **Dependencies** fixed when your values are not string type.
* **ExpressionChangedAfterItHasBeenCheckedError** fixed.

<a name="0.5.0"></a>
## [0.5.0](https://github.com/robust-team/angular-forms/compare/v0.4.0...v0.5.0) (2017-07-14)

### Changes

* Check Question refactored to CheckBox.
* Type for TextArea Question refactored of text-area to textarea.

<a name="0.4.0"></a>
## [0.4.0](https://github.com/robust-team/angular-forms/compare/v0.3.4...v0.4.0) (2017-07-12)

### Features

* New method in AngularFormsComponent (**isValid**, **getAnswersGroups** and **getAnswers**).

<a name="0.3.4"></a>
## [0.3.4](https://github.com/robust-team/angular-forms/compare/v0.3.3...v0.3.4) (2017-07-12)

### Features

* CSS classes in components (**rb-angular-forms**, **rb-fieldset** and **rb-data-table**) for facilitate customization.

### Bug Fixes

* Exception in DependencyService for unexisting dependencies.

<a name="0.3.3"></a>
## [0.3.3](https://github.com/robust-team/angular-forms/compare/v0.3.2...v0.3.3) (2017-07-12)

### Bug Fixes

* Validation messages' style.

<a name="0.3.2"></a>
## [0.3.2](https://github.com/robust-team/angular-forms/compare/v0.3.1...v0.3.2) (2017-07-10)

<a name="0.3.1"></a>
## [0.3.1](https://github.com/robust-team/angular-forms/compare/v0.3.0...v0.3.1) (2017-07-03)

<a name="0.3.0"></a>
# [0.3.0](https://github.com/robust-team/angular-forms/compare/v0.2.0...v0.3.0) (2017-07-03)

<a name="0.2.0"></a>
# 0.2.0 (2017-06-22)


demo/demo.service.ts
import { Injectable } from '@angular/core';

const groups: any[] = [
  {
    'code': 'G-01',
    'description': 'Fieldset Group',
    'type': 'group',
    'questions': [
      {
        'name': 'Q-0100',
        'description': 'Check Question 0100',
        'type': 'checkbox',
        'defaultOption': false,
        'validations': [
          { 'type': 'pattern', 'message': 'Required field.', 'value': 'true' }
        ]
      },
      {
        'name': 'Q-0101',
        'description': 'Radio Question 0101',
        'type': 'radio',
        'options': ['Option 1', 'Option 2'],
        'defaultOption': 'Option 1',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' }
        ],
        'disabled': true
      },
      {
        'name': 'Q-0102',
        'description': 'Select Question 0102',
        'type': 'select',
        'options': [
          { 'value': 'op-1',  'description': 'Option 1' },
          { 'value': 'op-2',  'description': 'Option 2' },
          { 'value': 'op-3',  'description': 'Option 3' }
        ],
        'editableOption': 'op-1',
        'placeholder': 'Select',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' }
        ],
        'answer': { 'value': 'op-2', 'id': 1 }
      },
      {
        'name': 'Q-0103',
        'description': 'Select Question 0103',
        'type': 'select',
        'options': ['Option 1', 'Option 2'],
        'editableOption': 'Option 2',
        'placeholder': 'Select',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' }
        ]
      },
      {
        'name': 'Q-0104',
        'description': 'Text Question 0104',
        'type': 'text',
        'answer': '200',
        'placeholder': 'Text Question 0104',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' },
          { 'type': 'minlength', 'message': 'Min Length: 3', 'value': 3 },
          { 'type': 'maxlength', 'message': 'Max Length: 10', 'value': 10 }
        ]
      },
      {
        'name': 'Q-0105',
        'description': 'Text Question 0105',
        'type': 'text',
        'placeholder': 'Text Question 0105',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' },
          { 'type': 'pattern', 'message': 'Format: 00000-000', 'value': '^\\d{5}-\\d{3}$' }
        ],
        'mask': '00000-000'
      },
      {
        'name': 'Q-0106',
        'description': 'TextArea Question 0106',
        'type': 'textarea',
        'placeholder': 'TextArea Question 0106',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' }
        ],
        'dependencies': [
          { 'code': 'Q-0103', 'criteria': 'equals', 'expectedAnswer': 'Option 1' }
        ]
      },
      {
        'name': 'Q-0107',
        'description': 'Text Question 0107',
        'type': 'text',
        'placeholder': 'Text Question 0107',
        'validations': [
          { 'type': 'required', 'message': 'Required field.' }
        ],
        'dependencies': [
          { 'code': 'Q-0104', 'criteria': 'lessthan', 'expectedAnswer': '300' },
          { 'code': 'Q-0104', 'criteria': 'notequals', 'expectedAnswer': '100' },
          { 'code': 'Q-0104', 'criteria': 'greaterthan', 'expectedAnswer': '100' },
          { 'code': 'Q-0104', 'criteria': 'lessthanorequals', 'expectedAnswer': '300' }
        ]
      },
      {
        'name': 'Q-0108',
        'description': 'Text Question 0108',
        'type': 'text',
        'placeholder': 'Text Question 0108',
        'validations': [
          { 'type': 'min', 'message': 'Min value: 1.', 'value': '1' },
          { 'type': 'max', 'message': 'Max value: 10.', 'value': '10' }
        ]
      },
      {
        'name': 'Q-0109',
        'description': 'Text Question 0109',
        'type': 'text',
        'placeholder': 'Text Question 0109',
        'validations': [
          { 'type': 'email', 'message': 'Invalid e-mail.' }
        ]
      }
    ]
  },
  {
    'code': 'G-02',
    'description': 'DataTable Group',
    'type': 'datatable',
    'validations': [
      { 'type': 'required', 'message': 'Min number of registers: 1.' },
      { 'type': 'minlength', 'message': 'Min number of registers: 1.', 'value': 1 },
      { 'type': 'maxlength', 'message': 'Max number of registers: 3.', 'value': 3 }
    ],
    'questions': [
      [
        {
          'name': 'Q-0201',
          'description': 'Select Question 0201',
          'type': 'select',
          'validations': [
            { 'type': 'required', 'message': 'Required field.' }
          ],
          'options': ['Option 1', 'Option 2', 'Option 3'],
          'editableOption': 'Option 3',
          'placeholder': 'Select'
        },
        {
          'name': 'Q-0202',
          'description': 'Text Question 0202',
          'type': 'text',
          'validations': [
            { 'type': 'required', 'message': 'Required field.' }
          ],
          'placeholder': 'Text Question 0202'
        }
      ],
      [
        {
          'name': 'Q-0201',
          'description': 'Select Question 0201',
          'type': 'select',
          'validations': [
            { 'type': 'required', 'message': 'Required field.' }
          ],
          'options': ['Option 1', 'Option 2', 'Option 3'],
          'placeholder': 'Select',
          'answer': { 'value': 'Option 2', 'id': 1 }
        },
        {
          'name': 'Q-0202',
          'description': 'Text Question 0202',
          'type': 'text',
          'validations': [
            { 'type': 'required', 'message': 'Required field.' }
          ],
          'placeholder': 'Text Question 0202',
          'answer': { 'value': 'Answer 0202', 'id': 1 }
        }
      ],
      [
        {
          'name': 'Q-0201',
          'description': 'Select Question 0201',
          'type': 'select',
          'validations': [
            { 'type': 'required', 'message': 'Required field.' }
          ],
          'options': ['Option 1', 'Option 2', 'Option 3'],
          'placeholder': 'Select',
          'answer': { 'value': 'Option 3', 'id': 2 }
        },
        {
          'name': 'Q-0202',
          'description': 'Text Question 0202',
          'type': 'text',
          'validations': [
            { 'type': 'required', 'message': 'Required field.' }
          ],
          'placeholder': 'Text Question 0202',
          'answer': { 'value': 'Answer 0202 2', 'id': 2 }
        }
      ]
    ]
  }
];

@Injectable()
export class DemoService {

  public getForm(): any[] {
    return groups;
  }
}
